pkg/controller/install/certresources.go

```diff
@@ -223,15 +231,74 @@ func (i *StrategyDeploymentInstaller) CertsRotated() bool {
 	return i.certificatesRotated
 }
 
-func ShouldRotateCerts(csv *v1alpha1.ClusterServiceVersion) bool {
+// shouldRotateCerts indicates whether an apiService cert should be rotated due to being
+// malformed, invalid, expired, inactive or within a specific freshness interval (DefaultCertMinFresh) before expiry.
+func shouldRotateCerts(certSecret *corev1.Secret, hosts []string) bool {
 	now := metav1.Now()
-	if !csv.Status.CertsRotateAt.IsZero() && csv.Status.CertsRotateAt.Before(&now) {
+	caPEM, ok := certSecret.Data[OLMCAPEMKey]
+	if !ok {
+		// missing CA cert in secret
+		return true
+	}
+	certPEM, ok := certSecret.Data["tls.crt"]
+	if !ok {
+		// missing cert in secret
+		return true
+	}
+
+	ca, err := certs.PEMToCert(caPEM)
+	if err != nil {
+		// malformed CA cert
+		return true
+	}
+	cert, err := certs.PEMToCert(certPEM)
+	if err != nil {
+		// malformed cert
 		return true
 	}
 
+	// check for cert freshness
+	if !certs.Active(ca) || now.Time.After(CalculateCertRotatesAt(ca.NotAfter)) ||
+		!certs.Active(cert) || now.Time.After(CalculateCertRotatesAt(cert.NotAfter)) {
+		return true
+	}
+
+	// Check validity of serving cert and if serving cert is trusted by the CA
+	for _, host := range hosts {
+		if err := certs.VerifyCert(ca, cert, host); err != nil {
+			return true
+		}
+	}
 	return false
 }
 
+func (i *StrategyDeploymentInstaller) ShouldRotateCerts(s Strategy) (bool, error) {
```

Comment:
(Similar to @tmshort's comment about existing code and reluctance to change it)...

It looks like this function is setup as an all or nothing thing? If _any_ of the certs needs to be rotated, we're going to rotate all of them?

In practice, maybe this doesn't actually make a difference because all of the certs are created initially at the same time with the same expiration. Thus generally all or none will need rotating?

Reply:
That's just the test to see if we need to transition the CSV from a Success/Failed state, there's a separate check before deciding to reuse/refresh certs for each deployment separately [here](https://github.com/operator-framework/operator-lifecycle-manager/pull/3217/files#diff-bc7b779225ba70ff0e1f63ff55cb2dd3d71b5ae4f96c158947266da9f8fde60dR381)

Reply:
Got it. Thanks!